File: LeetCode/Python Solutions/Top K Frequent Elements/frequent.py
from typing import List
# Solution
class Solution:
    def topKFrequent(self, nums: List[int], k: int) -> List[int]:
        Dict = {}
        for i in nums:
            if i not in Dict:
                Dict[i] = [1, i]
            else:
                Dict[i][0] += 1

        Sorted = sorted(Dict.values(), key=lambda x: x[0])[::-1]

        return [Sorted[k][1] for k in range(k)]


# Tests:
if __name__ == '__main__':
    Instant = Solution()
    Solve = Instant.topKFrequent(nums = [1,1,1,2,2,3], k = 2) 
    # nums = [1,1,1,2,2,3], k = 2 -> [1,2]
    print(Solve)

File: LeetCode/Java Solutions/Top K Frequent Elements/frequent.java
import java.util.*;

class Solution {
    public int[] topKFrequent(int[] nums, int k) {
        // Step 1: Create a frequency map
        Map<Integer, Integer> count = new HashMap<>();
        for (int num : nums) {
            count.put(num, count.getOrDefault(num, 0) + 1);
        }

        // Step 2: Build a heap with a custom comparator
        PriorityQueue<int[]> heap = new PriorityQueue<>(new Comparator<int[]>() {
            public int compare(int[] a, int[] b) {
                return b[0] - a[0];
            }
        });

        // Step 3: Add elements to the heap
        for (Map.Entry<Integer, Integer> entry : count.entrySet()) {
            heap.add(new int[]{entry.getValue(), entry.getKey()});
        }

        // Step 4: Extract the top k elements
        int[] result = new int[k];
        for (int i = 0; i < k; i++) {
            result[i] = heap.poll()[1];
        }
        return result;
    }
}

// Tests:
public class Main {
    public static void main(String[] args) {
        Solution solution = new Solution();
        int[] nums = {1,1,1,2,2,3};
        int k = 2;
        int[] result = solution.topKFrequent(nums, k);
        System.out.println(Arrays.toString(result));
    }
}
